### PR TITLE
Add tests for EC public key encoding/decoding with the legacy APIs

### PR DESCRIPTION
**Pull request by kaduk**
Add some tests that expose the issues reported at #14258 .  The issues are not fixed, so the CI is expected to fail.

##### Checklist

- [x] tests are added or updated

https://github.com/kaduk/openssl/commit/8d12b3bf581c02548d593c0852d538d17a509e8a is a crude hack that I used in order to get just the first commit's test to be able to pass, but I do not have even bad workarounds like that for the other two issues.

### CODE REVIEW

**Comment by kaduk** [comment]
@levitte IIRC you had your fingers on a lot of this stuff; please take a look

File: test/evp_extra_test.c [hunk 1]
```diff
@@ -2414,6 +2414,104 @@ static int test_evp_iv(int idx)
...
+    int ret = 0, len, savelen;
+    int nid;
+    unsigned char buf[1024];
+    unsigned char *p;
+    const unsigned char *q;
+    EVP_PKEY *pkey = NULL, *pkey2 = NULL, *pkeytmp = NULL;
+    EVP_PKEY_CTX *ctx = NULL;
+    EC_KEY *ec = NULL;
+
+    switch(idx) {
```
Comment (paulidale) [inline test/evp_extra_test.c:2428]: Why not put these NIDs into an array and use `nid = array[idx]`?  With a switch to `OSSL_NELEM` in the ADD_ALL_TESTS macro later.

Comment (kaduk) [reply test/evp_extra_test.c]: No reason at all...fixup pushed, and thanks for the suggestion

**Review by paulidale — APPROVED** [review]

**Comment by beldmit** [comment]
I'd prefer not to merge known-broken tests to the repo...

**Comment by paulidale** [comment]
I wasn't expecting broken tests to be merged.  Wait until they might pass.

**Comment by beldmit** [comment]
Sorry for misunderstanding

File: test/evp_extra_test.c [hunk 2]
```diff
@@ -2512,6 +2573,7 @@ int setup_tests(void)
     ADD_TEST(test_rand_agglomeration);
     ADD_ALL_TESTS(test_evp_iv, 10);
     ADD_TEST(test_EVP_rsa_pss_with_keygen_bits);
+    ADD_ALL_TESTS(test_ecpub, OSSL_NELEM(ecpub_nids));
```
Comment (levitte) [inline test/evp_extra_test.c:2576]: This, as well as `test_ecpub`, needs to be wrapped in a `#ifndef OPENSSL_NO_EC` guard.  From the CI builds:

```
In file included from test/evp_extra_test.c:33:0:
test/evp_extra_test.c: In function 'test_ecpub':
test/evp_extra_test.c:2456:32: error: implicit declaration of function 'EC_KEY_new_by_curve_name'; did you mean 'EVP_PKEY_get_group_name'? [-Werror=implicit-function-declaration]
             || !TEST_ptr((ec = EC_KEY_new_by_curve_name(nid)))
                                ^
test/testutil.h:449:64: note: in definition of macro 'TEST_ptr'
 # define TEST_ptr(a)          test_ptr(__FILE__, __LINE__, #a, a)
                                                                ^
test/evp_extra_test.c:2456:30: error: assignment makes pointer from integer without a cast [-Werror=int-conversion]
             || !TEST_ptr((ec = EC_KEY_new_by_curve_name(nid)))
                              ^
test/testutil.h:449:64: note: in definition of macro 'TEST_ptr'
 # define TEST_ptr(a)          test_ptr(__FILE__, __LINE__, #a, a)
                                                                ^
test/evp_extra_test.c:2457:27: error: implicit declaration of function 'EVP_PKEY_assign_EC_KEY'; did you mean 'EVP_PKEY_assign_RSA'? [-Werror=implicit-function-declaration]
             || !TEST_true(EVP_PKEY_assign_EC_KEY(pkeytmp, ec)))
Makefile:11327: recipe for target 'test/evp_extra_test-bin-evp_extra_test.o' failed
                           ^
test/testutil.h:462:66: note: in definition of macro 'TEST_true'
 # define TEST_true(a)         test_true(__FILE__, __LINE__, #a, (a) != 0)
                                                                  ^
test/evp_extra_test.c:2475:5: error: implicit declaration of function 'EC_KEY_free'; did you mean 'EVP_KEM_free'? [-Werror=implicit-function-declaration]
     EC_KEY_free(ec);
     ^~~~~~~~~~~
     EVP_KEM_free
cc1: all warnings being treated as errors
```

**Review by levitte — CHANGES_REQUESTED** [review]

File: test/evp_extra_test.c (hunk 2 above)
Comment (kaduk) [reply test/evp_extra_test.c:2576]: Thanks for spotting that; I will update shortly.

But while I'm happy to have review on the tests, I was hoping more to get your insight on the library code that is being tested (and found wanting)...

**Comment by levitte** [comment]
> I wasn't expecting broken tests to be merged. Wait until they might pass.

For all intents and purposes, this PR is approved and marked "approved: done!", which suggests that it *is* done and a candidate for merging.  If you expect further work to be done in this PR, then those approvals are quite confusing.

**Comment by levitte** [comment]
One thing is a bit faulty in the added test...  `pkey2` and `pkeytmp` will end up being the same `EVP_PKEY`, and yet, they are treated as separate `EVP_PKEY`s.  This diff fixes that problem.

``` diff
diff --git a/test/evp_extra_test.c b/test/evp_extra_test.c
index 72d32ed2d2..fff51c5fc8 100644
--- a/test/evp_extra_test.c
+++ b/test/evp_extra_test.c
@@ -2427,7 +2427,7 @@ static int test_ecpub(int idx)
     unsigned char buf[1024];
     unsigned char *p;
     const unsigned char *q;
-    EVP_PKEY *pkey = NULL, *pkey2 = NULL, *pkeytmp = NULL;
+    EVP_PKEY *pkey = NULL, *pkey2 = NULL;
     EVP_PKEY_CTX *ctx = NULL;
     EC_KEY *ec = NULL;
 
@@ -2452,14 +2452,13 @@ static int test_ecpub(int idx)
 
     /* Now try to decode the just-created DER. */
     q = buf;
-    if (!TEST_ptr((pkeytmp = EVP_PKEY_new()))
+    if (!TEST_ptr((pkey2 = EVP_PKEY_new()))
             || !TEST_ptr((ec = EC_KEY_new_by_curve_name(nid)))
-            || !TEST_true(EVP_PKEY_assign_EC_KEY(pkeytmp, ec)))
+            || !TEST_true(EVP_PKEY_assign_EC_KEY(pkey2, ec)))
         goto done;
     /* EC_KEY ownership transferred */
     ec = NULL;
-    if (!TEST_ptr((pkey2 = d2i_PublicKey(EVP_PKEY_EC, &pkeytmp, &q,
-                                                savelen))))
+    if (!TEST_ptr((d2i_PublicKey(EVP_PKEY_EC, &pkey2, &q, savelen))))
         goto done;
     /* The keys should match. */
     if (!TEST_int_eq(EVP_PKEY_cmp(pkey, pkey2), 1))
@@ -2470,7 +2469,6 @@ static int test_ecpub(int idx)
  done:
     EVP_PKEY_CTX_free(ctx);
     EVP_PKEY_free(pkey);
-    EVP_PKEY_free(pkeytmp);
     EVP_PKEY_free(pkey2);
     EC_KEY_free(ec);
     return ret;
```

I have identified the issues that this test hits, though.  PR coming up!

**Comment by levitte** [comment]
One of the issues is that we currently don't have a "type-specific" EC key->DER encoder for public keys (specifically!), and we don't have the corresponding "type-specific" DER->EC key decoder either.

A reason is that the old type specific functions are called `i2o_ECPublicKey()` and `o2i_ECPublicKey()`.  That name is much more indicative than one might think, as the blob that it encodes to / decodes from *is not DER*, it's just an octet (hence `o` rather than `d`) array.

So that begs the question, should we be very strict and invent some other output type "DER" for this type of stuff (that's the output type given to `OSSL_ENCODER_CTX_new_for_pkey()`, for example), or should we do like `i2d_PublicKey()` and pretend it's DER even thought it's absolutely not?
"OCT"?  "BLOB" (not the same as "MSBLOB", mind...)?

**Comment by t8m** [comment]
IMO i2d_PublicKey and d2i_PublicKey should simply fail on EVP_PKEYs with EC keys. Overloading the DER term to mean something completely different with EC keys does not seem right to me.

**Comment by t8m** [comment]
And yeah we have EVP_PKEY_set1_encoded_public_key() and EVP_PKEY_get1_encoded_public_key() for the cases where you need the encoded EC public key.

**Comment by levitte** [comment]
> IMO i2d_PublicKey and d2i_PublicKey should simply fail on EVP_PKEYs with EC keys

Strictly speaking, I fully agree, but that's more of an indication that i2d_PublicKey was perhaps not the best name choice. Alas, we have history, and what you propose would be a breaking change.

**Comment by t8m** [comment]
Then this should be probably special-cased in the actual i2d_PublicKey and d2i_PublicKey implementation. I would rather not try to invent another kind of encoder/decoder just for it as it would have to be a special case in those functions anyway.

**Comment by levitte** [comment]
> Then this should be probably special-cased in the actual i2d_PublicKey and d2i_PublicKey implementation

That's exactly how it is when the EVP_PKEY internal key is legacy

**Comment by kaduk** [comment]
> If you expect further work to be done in this PR, then those approvals are quite confusing.

In light of https://github.com/openssl/openssl/issues/14258#issuecomment-783351031 I was considering reworking the "d2i" side a bit to use the non-deprecated API.  But, the test file already defines `OPENSSL_SUPPRESS_DEPRECATED` so maybe it is not a pressing matter.

**Comment by levitte** [comment]
As can be seen #14291, I went for "blob".  "blob" is a fairly common term used for an unstructured bunch of bytes, that seemed appropriate enough.  I did consider "oct" for octet string, but thought that could be miscontrued as meaning something quite specific...  which it is for the moment, but who knows what will show up in the future?

**Comment by levitte** [comment]
This is being sucked into #14291.  I wonder if you want to clean this up into 1 commit, or shall I do so in #14291?

**Comment by kaduk** [comment]
> This is being sucked into #14291. I wonder if you want to clean this up into 1 commit, or shall I do so in #14291? If you do, I'll make sure to replace what I currently have with the cleaned up commit.

It seems simplest for you to just do it there, since then there isn't any need to synchronize changes.
(I mean, I'm happy to do it if that's easier for you for some reason, but it's not clear that it would really help.)

Do you think we should convert the d2i branch of the test into the non-deprecated way of getting a parameters (group)-only EVP_PKEY?

**Comment by levitte** [comment]
> (I mean, I'm happy to do it if that's easier for you for some reason, but it's not clear that it would really help.)

Different people are differently sensitive to how the commit message(s) get edited, i.e. how the commit is presented.  I don't know where you stand, so that was my giving you the opportunity to stay in control of that.  Sounds like you trust me to do a good job, though, so...

> Do you think we should convert the d2i branch of the test into the non-deprecated way of getting a parameters (group)-only EVP_PKEY?

It would hurt to *add* that, but I think it's also good to test the old stuff, just to make sure we don't mess that up.

**Comment by levitte** [comment]
This got merged as part of #14291, so closing it.  It worked very well as a demonstration to work from!

**Pull request closed without merging**